Update weekly refresh logic

quazi-h opened the pull request:
## Why
[Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1693)
We bounded the DAP extraction pulls to only pull 2020 records.

## This PR
Removed the endTime date bounds in the calls to the extraction pipelines in the manual refresh.
Added the monthly arms to the Environment filters and the annual 2021 arm for the CSLB extraction filters.
Fixed a bug that slipped in from a previous merge.

quazi-h committed:
Removed the endTime parameter from the extraction pipeline calls within the manual_refresh script.
This removes the righthand date bound on st_dap_pack_date, cslb_date, and bl_dap_pack_date for the hles, cslb, and environment pulls, respectively.
Added additional arms to the Environment and CSLB extraction pipeline filters.

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1693-update-weekly-refresh-logic

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1693-update-weekly-refresh-logic

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1693-update-weekly-refresh-logic

quazi-h committed:
typo

quazi-h committed:
error that slipped in from a previous commit (DAP bifurication
#149)

aherbst-broad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/cslb/CslbExtractionPipeline.scala:
```diff
@@ -30,7 +30,9 @@ object CslbExtractionPipeline extends ScioApp[Args] {
-  val arm = List("annual_2020_arm_1")
+
+  val arm =
+    List("annual_2020_arm_1", "annual_2021_arm_1")
```
What do you think about parameterizing these as cli args so we don't have to PR these sorts of changes going forward? I foresee a scenario where we'll be asked to pull the 2020_arm1 again for whatever reason and a full PR cycle would be cumbersome.

aherbst-broad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/environment/EnvironmentExtractionPipeline.scala:
```diff
@@ -60,7 +60,13 @@ object EnvironmentExtractionPipeline extends ScioApp[Args] {
       "nov2020_secondary_arm_1",
       "dec2020_arm_1",
-      "dec2020_secondary_arm_1"
+      "dec2020_secondary_arm_1",
```
Same comment about parameterizing these. I'm wondering if we can consume  "start month" + "end month" args and then dynamically create this list from there.

quazi-h committed:
Merge branch 'master' into QH-DSPDC-1693-update-weekly-refresh-logic

quazi-h committed:
Replaced the "arm" variable with a generator that will use provided args to compile a list of arms.
If args are not provided, we will use Jan 2020 as the default start date and the current time as the default end date.

aherbst-broad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/cslb/CslbExtractionPipeline.scala:
```diff
@@ -30,14 +32,35 @@ object CslbExtractionPipeline extends ScioApp[Args] {
+
+  // use args.startTime and args.endTime to determine full list
+  // branching logic: if args are provided, use dates to generate a list of the subset of arms
+  def extractionArmsGenerator(args: Args): List[String] = {
```
Looks good; let's get a unit test around this logic as well.

quazi-h committed:
Updated logic to generate list of years and list of monthYears in the Cslb and Environment extraction pipelines respectively.
Updated HLES and Sample extraction pipelines to pass through the initial HLES event to the pipeline builder correctly.
Updated the extraction pipeline builder and relevant integration and unit tests with changes to accommodate the code refactor.
Added unit tests for the Cslb and Environment extraction pipelines to test new logic and edge cases.

quazi-h committed:
Added an integration test for Environment that does not appear to be working.

aherbst-broad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/cslb/CslbExtractionPipeline.scala:
```diff
@@ -3,11 +3,16 @@ package org.broadinstitute.monster.dap.cslb
+
 // Ignore IntelliJ, this is used to make the implicit parser compile.
 import Args._
+class CslbExtractionFailException() extends scala.Exception {}
```
The parens and `scala.` are redundant here. I think you can do:

```suggestion
class CslbExtractionFailException extends Exception
```

aherbst-broad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/environment/EnvironmentExtractionPipeline.scala:
```diff
@@ -3,11 +3,20 @@ package org.broadinstitute.monster.dap.environment
 
+  val formatter = DateTimeFormatter.ofPattern("MMMyyyy")
+
+  val ENVEpoch = OffsetDateTime.of(2020, 1, 1, 0, 0, 0, 0, ZoneOffset.ofHours(-5))
```
⛏️ I think everywhere else we are using `Environment`, not `ENV`, I think this should be consistent with the rest of the naming in this file.

aherbst-broad commented on dap-etl/extraction/src/test/scala/org/broadinstitute/monster/dap/environment/EnvironmentExtractionPipelineBuilderIntegrationSpec.scala:
```diff
@@ -0,0 +1,73 @@
+
+  behavior of "EnvironmentSurveyExtractionPipelineBuilder"
+
+  it should "successfully download records from RedCap" in {
```
❤️ I love that there is now a test for this.

aherbst-broad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/sample/SampleExtractionPipeline.scala:
```diff
@@ -31,14 +31,13 @@ object SampleExtractionPipeline extends ScioApp[Args] {
       forms,
       extractionFiltersGenerator,
-      arm,
+      (_, _) => List("baseline_arm_1"),
```
Same comment about keeping the constant for use in specs:

```suggestion
      (_, _) => List(arm),
```

aherbst-broad commented on dap-etl/extraction/src/test/scala/org/broadinstitute/monster/dap/hles/HLESExtractionPipelineBuilderIntegrationSpec.scala:
```diff
@@ -33,7 +33,7 @@ class HLESExtractionPipelineBuilderIntegrationSpec extends PipelineBuilderSpec[A
   val studyIdsRequest = RedcapRequestGeneratorParams(
     testArgs.apiToken,
-    HLESurveyExtractionPipeline.arm,
+    List("baseline_arm_1"),
```
🔧 If we keep the constant in the related file, this would become:
```suggestion
    List(HLESurveyExtractionPipeline.arm),
```

aherbst-broad reviewed (commented):
lgtm, I will approve once tests are passing.

quazi-h committed:
Updated logic in ExtractionPipelineBuilder to dispatch requests for the list of records for all arms instead of doing one request per arm.
Made "arms" a variable in the integration tests + general pipeline builder script.

quazi-broad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/cslb/CslbExtractionPipeline.scala:
```diff
@@ -30,14 +35,38 @@ object CslbExtractionPipeline extends ScioApp[Args] {
+    if (start.isAfter(end)) throw new CslbExtractionFailException
+    val dateList =
+      (Iterator.iterate(start)(_ plusDays 1) takeWhile (_ isBefore end.plusDays(1))).toList
+    dateList.map(date => date.getYear).distinct
```
I was originally using "toSet.toList" but switched to using "distinct" because by default, the "Set" interface doesn't preserve element order - caught that from one of the unit tests

raaidbroad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/environment/EnvironmentExtractionPipeline.scala:
```diff
@@ -3,11 +3,20 @@ package org.broadinstitute.monster.dap.environment
 // Ignore IntelliJ, this is used to make the implicit parser compile.
 import Args._
 
+class EnvironmentExtractionFailException() extends scala.Exception {}
```
same thing as Drew's comment earlier, should be able to just use `Exception`

raaidbroad commented on dap-etl/extraction/src/main/scala/org/broadinstitute/monster/dap/environment/EnvironmentExtractionPipeline.scala:
```diff
@@ -3,11 +3,20 @@ package org.broadinstitute.monster.dap.environment
+
 object EnvironmentExtractionPipeline extends ScioApp[Args] {
 
+  val formatter = DateTimeFormatter.ofPattern("MMMyyyy")
```
Should there be three `M`s here or just two? 👀

raaidbroad reviewed (approved):
Barring a few small tweaks that aren't crazy important (except possibly the date format?) looks p good to me!

quazi-h committed:
Removed some redundant code, updated some variables for consistency.
Refactored code to employ the use of constants for ease of use.

quazi-h committed:
Added input arguments to the manual_refresh.sh script to allow passing in the start and end variables as needed
